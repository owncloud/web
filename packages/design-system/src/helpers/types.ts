export interface ContextualHelperDataListItem {
  text: string
  headline?: boolean
}

export interface ContextualHelperData {
  title?: string
  text?: string
  list?: ContextualHelperDataListItem[]
  readMoreLink?: string
}

export interface ContextualHelper {
  isEnabled: boolean
  data: ContextualHelperData
}

export interface PasswordPolicyRule {
  code: string
  message: string
  helperMessage?: string
  format: (number | string)[]
  verified: boolean
}

export interface PasswordPolicy {
  rules: unknown[]

  check(password: string): boolean

  missing(password: string): {
    rules: PasswordPolicyRule[]
  }
}
