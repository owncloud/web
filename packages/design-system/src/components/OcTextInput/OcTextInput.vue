<template>
  <div :class="$attrs.class">
    <slot name="label">
      <label class="oc-label" :for="id" v-text="label" />
    </slot>
    <div class="oc-position-relative">
      <oc-icon
        v-if="readOnly"
        name="lock"
        size="small"
        class="oc-mt-s oc-ml-s oc-position-absolute"
      />
      <component
        :is="inputComponent"
        :id="id"
        v-bind="additionalAttributes"
        ref="input"
        :aria-invalid="ariaInvalid"
        class="oc-text-input oc-input oc-rounded"
        :class="{
          'oc-text-input-warning': !!warningMessage,
          'oc-text-input-danger': !!errorMessage,
          'oc-pl-l': !!readOnly
        }"
        :type="type"
        :value="displayValue"
        :disabled="disabled || readOnly"
        v-on="additionalListeners"
        @change="onChange(($event.target as HTMLInputElement).value)"
        @input="onInput(($event.target as HTMLInputElement).value)"
        @password-challenge-completed="$emit('passwordChallengeCompleted')"
        @password-challenge-failed="$emit('passwordChallengeFailed')"
        @focus="onFocus($event.target)"
      />
      <oc-button
        v-if="showClearButton"
        :aria-label="clearButtonAccessibleLabelValue"
        class="oc-pr-s oc-position-center-right oc-text-input-btn-clear"
        appearance="raw"
        @click="onClear"
      >
        <oc-icon name="close" size="small" variation="passive" />
      </oc-button>
    </div>
    <div
      v-if="showMessageLine"
      class="oc-text-input-message"
      :class="{
        'oc-text-input-description': !!descriptionMessage,
        'oc-text-input-warning': !!warningMessage,
        'oc-text-input-danger': !!errorMessage
      }"
    >
      <oc-icon
        v-if="messageText !== null && !!descriptionMessage"
        name="information"
        size="small"
        fill-type="line"
        accessible-label="info"
        aria-hidden="true"
      />

      <span
        :id="messageId"
        :class="{
          'oc-text-input-description': !!descriptionMessage,
          'oc-text-input-warning': !!warningMessage,
          'oc-text-input-danger': !!errorMessage
        }"
        v-text="messageText"
      />
    </div>
    <portal-target v-if="type === 'password'" name="app.design-system.password-policy" />
  </div>
</template>

<script lang="ts">
import { defineComponent, HTMLAttributes, PropType } from 'vue'

import uniqueId from '../../utils/uniqueId'
import OcButton from '../OcButton/OcButton.vue'
import OcIcon from '../OcIcon/OcIcon.vue'
import OcTextInputPassword from '../_OcTextInputPassword/_OcTextInputPassword.vue'
import { PasswordPolicy } from '../../helpers'

/**
 * Form Inputs are used to allow users to provide text input when the expected
 * input is short. Form Input has a range of options and supports several text
 * formats including numbers. For longer input, use the form `Textarea` element.*
 *
 * ## Accessibility
 * The label is required and represents the name of the input.
 *
 * The description-message can be used additionally to give further information about the input field. When a
 * description is given, it will be automatically referenced via the `aria-describedby` property.
 * An error or warning will replace the description as well as the `aria-describedby` property until the error
 * or warning is fixed.
 */
export default defineComponent({
  name: 'OcTextInput',
  components: { OcIcon, OcButton, OcTextInputPassword },
  status: 'ready',
  release: '1.0.0',
  inheritAttrs: false,
  props: {
    /**
     * The ID of the element.
     */
    id: {
      type: String,
      required: false,
      default: () => uniqueId('oc-textinput-')
    },
    /**
     * The type of the form input field.
     * `text, number, email, password`
     */
    type: {
      type: String,
      default: 'text',
      validator: (value: string) => {
        return ['text', 'number', 'email', 'password'].includes(value)
      }
    },
    /**
     * Text value of the form input field.
     * @model
     */
    modelValue: {
      type: String,
      required: false,
      default: null
    },
    /**
     * Selection range to accomplish partial selection
     */
    selectionRange: {
      type: Array as unknown as PropType<[number, number]>,
      required: false,
      default: null
    },
    /**
     * Whether or not the input element should have a dedicated button for clearing the input content.
     */
    clearButtonEnabled: {
      type: Boolean,
      required: false,
      default: false
    },
    /**
     * The aria label for the clear button. Only used if it's enabled at all.
     */
    clearButtonAccessibleLabel: {
      type: String,
      required: false,
      default: ''
    },
    /**
     * Value to show when no value is provided
     * This does not set `value` automatically.
     * The user needs to explicitly enter a text to set it as `value`.
     */
    defaultValue: {
      type: String,
      required: false,
      default: null
    },
    /**
     * Disables the input field
     */
    disabled: {
      type: Boolean,
      default: false
    },
    /**
     * Accessible label of the form input field.
     **/
    label: {
      type: String,
      required: true,
      default: null
    },
    /**
     * A warning message which is shown below the input.
     */
    warningMessage: {
      type: String,
      default: null
    },
    /**
     * An error message which is shown below the input.
     */
    errorMessage: {
      type: String,
      default: null
    },
    /**
     * Whether or not vertical space below the input should be reserved for a one line message,
     * so that content actually appearing there doesn't shift the layout.
     */
    fixMessageLine: {
      type: Boolean,
      default: false
    },
    /**
     * A description text which is shown below the input field.
     */
    descriptionMessage: {
      type: String,
      default: null
    },
    /**
     * Determines if the input field is read only.
     *
     * Read only field will be visualized by a lock item and additionally behaves like a disabled field.
     * Read only takes effect if the server won't allow to change the value at all,
     * disabled should be used instead, if the value can't be changed in a specific context.
     *
     * For example: If the backend doesn't allow to set the login states for users in general, use read only.
     * If it's not allowed to change for the current logged-in User, use disabled.
     *
     */
    readOnly: {
      type: Boolean,
      default: false
    },
    /**
     * Array of password policy rules, if type is password and password policy is given,
     * the entered value will be checked against these rules.
     *
     * Password policy rules must be compliant with auth0/password-sheriff
     * https://github.com/auth0/password-sheriff
     *
     */
    passwordPolicy: {
      type: Object as PropType<PasswordPolicy>,
      default: () => ({})
    },
    /**
     * Method to generate random password
     */
    generatePasswordMethod: {
      type: Function as PropType<(...args: unknown[]) => string>,
      required: false,
      default: null
    }
  },
  emits: [
    'change',
    'update:modelValue',
    'focus',
    'passwordChallengeCompleted',
    'passwordChallengeFailed'
  ],
  computed: {
    showMessageLine() {
      return (
        this.fixMessageLine ||
        !!this.warningMessage ||
        !!this.errorMessage ||
        !!this.descriptionMessage
      )
    },
    messageId() {
      return `${this.id}-message`
    },
    additionalListeners() {
      if (this.type === 'password') {
        return { passwordGenerated: this.onInput }
      }

      return {}
    },
    additionalAttributes() {
      const additionalAttrs: Record<string, unknown> = {}
      if (!!this.warningMessage || !!this.errorMessage || !!this.descriptionMessage) {
        additionalAttrs['aria-describedby'] = this.messageId
      }
      // FIXME: placeholder usage is discouraged, we need to find a better UX concept
      if (this.defaultValue) {
        additionalAttrs['placeholder'] = this.defaultValue
      }
      if (this.type === 'password') {
        additionalAttrs['password-policy'] = this.passwordPolicy
        additionalAttrs['generate-password-method'] = this.generatePasswordMethod
        additionalAttrs['has-warning'] = !!this.warningMessage
        additionalAttrs['has-error'] = !!this.errorMessage
      }
      // Exclude listeners for events which are handled via methods in this component
      // eslint-disable-next-line no-unused-vars
      const { change, input, focus, class: classes, ...attrs } = this.$attrs

      return { ...attrs, ...additionalAttrs }
    },
    ariaInvalid() {
      return (!!this.errorMessage).toString() as HTMLAttributes['aria-invalid']
    },
    messageText() {
      if (this.errorMessage) {
        return this.errorMessage
      }

      if (this.warningMessage) {
        return this.warningMessage
      }

      return this.descriptionMessage
    },
    showClearButton() {
      return !this.disabled && this.clearButtonEnabled && !!this.modelValue
    },
    clearButtonAccessibleLabelValue() {
      return this.clearButtonAccessibleLabel || this.$gettext('Clear input')
    },
    displayValue() {
      return this.modelValue || ''
    },
    inputComponent() {
      return this.type === 'password' ? 'oc-text-input-password' : 'input'
    }
  },
  methods: {
    /**
     * Puts focus on this input element
     * @public
     */
    focus() {
      ;(this.$refs.input as HTMLInputElement).focus()
    },
    onClear() {
      this.focus()

      this.onInput(null)
      this.onChange(null)
    },
    onChange(value: string) {
      /**
       * Change event
       * @type {event}
       **/
      this.$emit('change', value)
    },
    onInput(value: string) {
      /**
       * Input event
       * @type {event}
       **/
      this.$emit('update:modelValue', value)
    },
    onFocus(target: HTMLInputElement) {
      target.select()
      if (this.selectionRange && this.selectionRange.length > 1) {
        target.setSelectionRange(this.selectionRange[0], this.selectionRange[1])
      }
      /**
       * Focus event - emitted as soon as the input field is focused
       * @type {event}
       **/
      this.$emit('focus', target.value)
    }
  }
})
</script>

<style lang="scss">
.oc-text-input-message.oc-text-input-description {
  display: flex;
  align-items: center;
  position: relative;
  padding-left: var(--oc-space-large);
  padding-top: calc(var(--oc-space-xsmall) - 2px);

  .oc-icon {
    position: absolute;
    left: var(--oc-space-xsmall);
    top: var(--oc-space-xsmall);
  }
}

.oc-text-input {
  &-description {
    color: var(--oc-color-text-muted);
  }

  &-success,
  &-success:focus {
    border-color: var(--oc-color-swatch-success-default) !important;
    color: var(--oc-color-swatch-success-default) !important;
  }

  &-warning,
  &-warning:focus {
    border-color: var(--oc-color-swatch-warning-default) !important;
    color: var(--oc-color-swatch-warning-default) !important;
  }

  &-danger,
  &-danger:focus {
    border-color: var(--oc-color-swatch-danger-default) !important;
    color: var(--oc-color-swatch-danger-default) !important;
  }

  &-message {
    display: flex;
    align-items: center;
    margin-top: var(--oc-space-xsmall);
    min-height: $oc-font-size-default * 1.5;
  }
}
</style>

<docs>
```js
<template>
  <section>
    <h3 class="oc-heading-divider">
      Input Types
    </h3>
    <oc-text-input class="oc-mb-s" label="Text"/>
    <oc-text-input class="oc-mb-s" disabled label="Disabled" value="I am disabled"/>
    <oc-text-input class="oc-mb-s" read-only="true" label="Read only" value="I am read only"/>
    <oc-text-input class="oc-mb-s" type="number" label="Number"/>
    <oc-text-input class="oc-mb-s" type="email" label="Email"/>
    <oc-text-input class="oc-mb-s" type="password" label="Password"/>
    <h3 class="oc-heading-divider">
      Binding
    </h3>
    <oc-text-input label="Text" v-model="inputValue"/>
    <oc-text-input disabled label="Text" v-model="inputValue"/>
    <h3 class="oc-heading-divider">
      Interactions
    </h3>
    <oc-button @click="_focus" class="oc-my-m">Focus input below</oc-button>
    <oc-text-input label="Focus field" ref="inputForFocus"/>
    <oc-button @click="_focusAndSelect" class="oc-my-m">Focus and select input below</oc-button>
    <oc-text-input label="Select field" value="Will you select this existing text?" ref="inputForFocusSelect"/>
    <oc-text-input label="Clear input" v-model="inputValueForClearing" :clear-button-enabled="true"/>
    <oc-text-input label="Input with default" v-model="inputValueWithDefault" :clear-button-enabled="true"
                   default-value="Some default"/>
    <p>
      Value: {{ inputValueWithDefault !== null ? inputValueWithDefault : "null" }}
    </p>
    <h3 class="oc-heading-divider">
      Messages
    </h3>
    <oc-text-input
      label="Input with description message below"
      class="oc-mb-s"
      description-message="This is a description message."
      :fix-message-line="true"
    />
    <oc-text-input
      label="Input with error and warning messages with reserved space below"
      class="oc-mb-s"
      v-model="valueForMessages"
      :error-message="errorMessage"
      :warning-message="warningMessage"
      :fix-message-line="true"
    />
    <oc-text-input
      label="Input with error and warning messages without reserved space below"
      class="oc-mb-s"
      v-model="valueForMessages"
      :error-message="errorMessage"
      :warning-message="warningMessage"
    />
  </section>
</template>
<script>
  export default {
    data: () => {
      return {
        inputValue: 'initial',
        valueForMessages: '',
        inputValueForClearing: 'clear me',
        inputValueWithDefault: null,
      }
    },
    computed: {
      errorMessage() {
        return this.valueForMessages.length === 0 ? 'Value is required.' : ''
      },
      warningMessage() {
        return this.valueForMessages.endsWith(' ') ? 'Trailing whitespace should be avoided.' : ''
      }
    },
    methods: {
      _focus() {
        this.$refs.inputForFocus.focus()
      },
      _focusAndSelect() {
        this.$refs.inputForFocusSelect.focus()
      }
    }
  }
</script>
```
</docs>
