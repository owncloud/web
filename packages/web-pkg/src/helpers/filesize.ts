import { filesize, FileSizeOptionsString } from 'filesize'
import { getLocaleFromLanguage } from './locale'

const mb = 1048576

/**
 * Returns formatted size
 *
 * @param {Number|String} size Unformatted size
 * @param {String} currentLanguage
 * @returns {String} formatted size
 */
export const formatFileSize = (size: number | string, currentLanguage: string) => {
  const parsedSize = typeof size === 'string' ? parseInt(size) : size
  if (parsedSize < 0) {
    return '--'
  }

  if (isNaN(parsedSize)) {
    return '?'
  }

  return filesize<FileSizeOptionsString>(parsedSize, {
    round: parsedSize < mb ? 0 : 1,
    locale: getLocaleFromLanguage(currentLanguage),
    base: 10,
    output: 'string'
  })
}
