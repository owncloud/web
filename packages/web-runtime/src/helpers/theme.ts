import { isEqual } from 'lodash-es'
import defaultTheme from '../../themes/owncloud/theme.json'
import { v4 as uuidV4 } from 'uuid'
import { ThemingConfig } from '@ownclouders/web-pkg'

export const loadTheme = async (location = '') => {
  const defaultOwnCloudTheme = {
    defaults: {
      ...defaultTheme.clients.web.defaults,
      common: defaultTheme.common
    },
    themes: defaultTheme.clients.web.themes
  }

  if (location.split('.').pop() !== 'json') {
    if (isEqual(process.env.NODE_ENV, 'development')) {
      console.info(`Theme '${location}' does not specify a json file, using default theme.`)
    }
    return defaultOwnCloudTheme
  }

  try {
    const response = await fetch(location, { headers: { 'X-Request-ID': uuidV4() } })
    if (!response.ok) {
      console.error(`Failed to load theme '${location}', invalid response. Using default theme.`)
      return defaultOwnCloudTheme
    }

    const theme = await response.json()

    try {
      const parsedTheme = ThemingConfig.parse(theme)

      return {
        defaults: {
          common: parsedTheme.common,
          ...parsedTheme.clients.web.defaults
        },
        themes: parsedTheme.clients.web.themes
      }
    } catch (error) {
      console.error(
        `Failed to load theme '${location}', invalid theme. Using default theme.`,
        error
      )
      return defaultOwnCloudTheme
    }
  } catch {
    console.error(`Failed to load theme '${location}', using default theme.`)
    return defaultOwnCloudTheme
  }
}
