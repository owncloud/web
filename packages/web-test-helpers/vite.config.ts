import { resolve } from 'path'
import { defineConfig } from 'vite'
import dts from 'vite-plugin-dts'
import pkg from './package.json' assert { type: 'json' }
import vue from '@vitejs/plugin-vue'

export default defineConfig({
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-test-helpers',
      fileName: (format) => `web-test-helpers.${format}.js`
    },
    rollupOptions: {
      external: [...Object.keys(pkg.dependencies), ...Object.keys(pkg.peerDependencies)]
    }
  },
  plugins: [vue(), dts({ include: ['src'], outDir: 'dist/types', insertTypesEntry: true })]
})
