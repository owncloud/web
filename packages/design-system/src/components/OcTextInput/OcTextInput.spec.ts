import { shallowMount, mount, defaultPlugins } from 'web-test-helpers'
import OcTextInput from './OcTextInput.vue'
import { PasswordPolicy } from '../../helpers'
import { mock } from 'vitest-mock-extended'

const defaultProps = {
  label: 'label'
}

Object.assign(navigator, {
  clipboard: {
    writeText: vi.fn(),
    readText: vi.fn()
  }
})

// @vitest-environment jsdom
describe('OcTextInput', () => {
  function getShallowWrapper(props = {}) {
    return shallowMount(OcTextInput, {
      props: {
        ...defaultProps,
        ...props
      },
      global: { plugins: [...defaultPlugins()] }
    })
  }

  function getMountedWrapper(
    options: { props?: Record<string, unknown>; attachTo?: HTMLElement } = { props: {} },
    passwordPolicy = { active: false, pass: false }
  ) {
    const passwordPolicyMock = mock<PasswordPolicy>()
    passwordPolicyMock.missing.mockReturnValue({
      rules: [
        {
          code: 'minLength',
          message: '%{param1}+ letters',
          format: ['8'],
          verified: passwordPolicy.pass
        }
      ]
    })
    passwordPolicyMock.check.mockReturnValueOnce(passwordPolicy.pass)

    if (passwordPolicy.active) {
      options.props = { ...(options.props || {}), passwordPolicy: passwordPolicyMock }
    }

    return mount(OcTextInput, {
      ...options,
      global: {
        plugins: [...defaultPlugins()]
      }
    })
  }

  const selectors = {
    textInputMessage: '.oc-text-input-message span',
    clearInputButton: '.oc-text-input-btn-clear',
    inputField: '.oc-text-input',
    infoIcon: '.oc-text-input-message .oc-icon',
    showPasswordToggleBtn: '.oc-text-input-show-password-toggle',
    copyPasswordBtn: '.oc-text-input-copy-password-button',
    generatePasswordBtn: '.oc-text-input-generate-password-button'
  }

  describe('id prop', () => {
    const wrapper = getShallowWrapper({ id: 'test-input-id', descriptionMessage: 'hello' })
    it('should set provided id to the input', () => {
      expect(wrapper.find('input').attributes().id).toBe('test-input-id')
    })
    it('should set label target for provided id', () => {
      expect(wrapper.find('label').attributes().for).toBe('test-input-id')
    })
    it('should set message id according to provided id', () => {
      expect(wrapper.find(selectors.textInputMessage).attributes().id).toBe('test-input-id-message')
    })
  })

  describe('label prop', () => {
    it('should set provided label to the input', () => {
      const wrapper = getShallowWrapper()
      expect(wrapper.find('label').text()).toBe('label')
    })
  })

  describe('password input field', () => {
    describe('copy password button', () => {
      it('should not exist if type is not "password" or no value entered', () => {
        const wrapper = getMountedWrapper()
        expect(wrapper.find(selectors.copyPasswordBtn).exists()).toBeFalsy()
        const wrapper2 = getMountedWrapper({ props: { type: 'password' } })
        expect(wrapper2.find(selectors.copyPasswordBtn).exists()).toBeFalsy()
      })
      it('should not exist if the input is disabled', () => {
        const wrapper = getMountedWrapper({ props: { type: 'password', disabled: true } })
        expect(wrapper.find(selectors.copyPasswordBtn).exists()).toBeFalsy()
      })
      it('should exist if type is "password" and value entered', async () => {
        const wrapper = getMountedWrapper({ props: { type: 'password' } })
        await wrapper.find(selectors.inputField).setValue('password')
        expect(wrapper.find(selectors.copyPasswordBtn).exists()).toBeTruthy()
      })
      it('should copy password to clipboard if clicked', async () => {
        const wrapper = getMountedWrapper({ props: { type: 'password' } })
        await wrapper.find(selectors.inputField).setValue('password')
        await wrapper.find(selectors.copyPasswordBtn).trigger('click')
        expect(navigator.clipboard.writeText).toHaveBeenCalledWith('password')
      })
    })
    describe('show hide password toggle button', () => {
      it('should not exist if type is not "password" or no value entered', () => {
        const wrapper = getMountedWrapper()
        expect(wrapper.find(selectors.showPasswordToggleBtn).exists()).toBeFalsy()

        const wrapper2 = getMountedWrapper({ props: { type: 'password' } })
        expect(wrapper2.find(selectors.showPasswordToggleBtn).exists()).toBeFalsy()
      })
      it('should not exist if the input is disabled', () => {
        const wrapper = getMountedWrapper({ props: { type: 'password', disabled: true } })
        expect(wrapper.find(selectors.showPasswordToggleBtn).exists()).toBeFalsy()
      })
      it('should exist if type is "password" and value entered', async () => {
        const wrapper = getMountedWrapper({ props: { type: 'password' } })
        await wrapper.find(selectors.inputField).setValue('password')
        expect(wrapper.find(selectors.showPasswordToggleBtn).exists()).toBeTruthy()
      })
      it('should show password plaintext/veiled if clicked', async () => {
        const wrapper = getMountedWrapper({ props: { type: 'password' } })
        await wrapper.find(selectors.inputField).setValue('password')
        await wrapper.find(selectors.showPasswordToggleBtn).trigger('click')
        expect(wrapper.find(selectors.inputField).attributes().type).toBe('text')
        await wrapper.find(selectors.showPasswordToggleBtn).trigger('click')
        expect(wrapper.find(selectors.inputField).attributes().type).toBe('password')
      })
    })
    describe('generate password button', () => {
      it('should not exist if type is not "password" or prop "generatePasswordMethod" is not provided', () => {
        const wrapper = getMountedWrapper()
        expect(wrapper.find(selectors.generatePasswordBtn).exists()).toBeFalsy()

        const wrapper2 = getMountedWrapper({ props: { type: 'password' } })
        expect(wrapper2.find(selectors.generatePasswordBtn).exists()).toBeFalsy()
      })
      it('should not exist if the input is disabled', () => {
        const wrapper = getMountedWrapper({ props: { type: 'password', disabled: true } })
        expect(wrapper.find(selectors.generatePasswordBtn).exists()).toBeFalsy()
      })
      it('should exist if type is "password" and prop "generatePasswordMethod" is provided', () => {
        const wrapper = getMountedWrapper({
          props: { generatePasswordMethod: vi.fn(), type: 'password' }
        })
        expect(wrapper.find(selectors.generatePasswordBtn).exists()).toBeTruthy()
      })
      it('should fill input with generated password if clicked', async () => {
        const wrapper = getMountedWrapper({
          props: { generatePasswordMethod: vi.fn(() => 'PAssword12#!'), type: 'password' }
        })
        await wrapper.find(selectors.generatePasswordBtn).trigger('click')
        expect((wrapper.find(selectors.inputField).element as HTMLInputElement).value).toEqual(
          'PAssword12#!'
        )
      })
    })
    describe('password policy', () => {
      it('should emit "passwordChallengeFailed" if password does not match criteria', async () => {
        const wrapper = getMountedWrapper(
          {
            props: { type: 'password' }
          },
          { active: true, pass: false }
        )
        await wrapper.find(selectors.inputField).setValue('pass')
        expect(wrapper.emitted('passwordChallengeCompleted')).toBeFalsy()
      })
      it('should emit "passwordChallengeCompleted" if password matches criteria', async () => {
        const wrapper = getMountedWrapper(
          {
            props: { type: 'password' }
          },
          { active: true, pass: true }
        )
        await wrapper.find(selectors.inputField).setValue('password123')
        expect(wrapper.emitted('passwordChallengeCompleted')).toBeTruthy()
      })
      it('displays error state if password does not match criteria', async () => {
        const wrapper = getMountedWrapper(
          {
            props: { type: 'password' }
          },
          { active: true, pass: false }
        )
        await wrapper.find(selectors.inputField).setValue('pass')
        expect(wrapper.html()).toMatchSnapshot()
      })
      it('displays success state if password matches criteria', async () => {
        const wrapper = getMountedWrapper(
          {
            props: { type: 'password' }
          },
          { active: true, pass: true }
        )
        await wrapper.find(selectors.inputField).setValue('password123')
        expect(wrapper.html()).toMatchSnapshot()
      })
    })
  })

  describe('when a description message is provided', () => {
    const wrapper = getShallowWrapper({ descriptionMessage: 'You should pass.' })
    it('should add the description class to the input message', () => {
      expect(wrapper.find(selectors.textInputMessage).attributes().class).toContain(
        'oc-text-input-description'
      )
    })
    it('should show the description message as the input message text', () => {
      expect(wrapper.find(selectors.textInputMessage).text()).toBe('You should pass.')
    })

    const mountedWrapper = getMountedWrapper({ props: { descriptionMessage: 'You should pass.' } })
    it('should show an info icon', () => {
      expect(mountedWrapper.find(selectors.infoIcon).exists()).toBe(true)
    })
  })

  describe('when a warning message is provided', () => {
    const wrapper = getShallowWrapper({ warningMessage: 'You may pass.' })
    it('should add the warning class to the input', () => {
      expect(wrapper.find('input').attributes().class).toContain('oc-text-input-warning')
    })
    it('should add the warning class to the input message', () => {
      expect(wrapper.find(selectors.textInputMessage).attributes().class).toContain(
        'oc-text-input-warning'
      )
    })
    it('should show the warning message as the input message text', () => {
      expect(wrapper.find(selectors.textInputMessage).text()).toBe('You may pass.')
    })
  })

  describe('when an error message is provided', () => {
    const wrapper = getShallowWrapper({ errorMessage: 'You shall not pass.' })
    it('should add the error class to the input', () => {
      expect(wrapper.find('input').attributes().class).toContain('oc-text-input-danger')
    })
    it('should add the error class to the input message', () => {
      expect(wrapper.find(selectors.textInputMessage).attributes().class).toContain(
        'oc-text-input-danger'
      )
    })
    it('should show the error message as the input message text', () => {
      expect(wrapper.find(selectors.textInputMessage).text()).toBe('You shall not pass.')
    })
    it('should set the input aria-invalid attribute to true', () => {
      expect(wrapper.find('input').attributes('aria-invalid')).toBe('true')
    })
  })

  describe('message priority', () => {
    it('should give error message top priority', () => {
      const wrapper = getShallowWrapper({
        errorMessage: 'You shall not pass.',
        warningMessage: 'You may pass.',
        descriptionMessage: 'Your should pass.'
      })
      const messageEl = wrapper.find('.oc-text-input-message span')
      expect(messageEl.attributes().class).toBe(
        'oc-text-input-description oc-text-input-warning oc-text-input-danger'
      )
      expect(messageEl.text()).toBe('You shall not pass.')
    })
    it('should give waring message priority over description message', () => {
      const wrapper = getShallowWrapper({
        warningMessage: 'You may pass.',
        descriptionMessage: 'Your should pass.'
      })
      const messageEl = wrapper.find(selectors.textInputMessage)
      expect(messageEl.attributes().class).toBe('oc-text-input-description oc-text-input-warning')
      expect(messageEl.text()).toBe('You may pass.')
    })
  })

  describe('type prop', () => {
    it('should only allow text, number, email and password as type', () => {
      expect(OcTextInput.props.type.validator('binary')).toBeFalsy()
    })
    it.each(['text', 'number', 'email', 'password'])(
      'should set the provided type for the input',
      (type) => {
        const wrapper = getMountedWrapper({ props: { type: type } })
        expect(wrapper.find('input').attributes('type')).toBe(type)
      }
    )
  })

  describe('input events', () => {
    it('should emit an input event on typing', async () => {
      const wrapper = getShallowWrapper()
      expect(wrapper.emitted().input).toBeFalsy()
      await wrapper.find('input').setValue('a')
      expect(wrapper.emitted('update:modelValue')).toBeTruthy()
      expect(wrapper.emitted('update:modelValue')[0][0]).toBe('a')
    })
  })

  describe('clear-button accessible label prop', () => {
    it('should set the aria label attribute if provided', () => {
      const wrapper = getShallowWrapper({
        clearButtonEnabled: true,
        modelValue: 'non-empty-value',
        clearButtonAccessibleLabel: 'test label'
      })
      expect(wrapper.find(selectors.clearInputButton).attributes().arialabel).toBe('test label')
    })
  })

  describe('clear input', () => {
    it('has no clear button when it is disabled', () => {
      const wrapper = getShallowWrapper({
        modelValue: 'non-empty-value',
        clearButtonEnabled: true,
        disabled: true
      })
      expect(wrapper.find(selectors.clearInputButton).exists()).toBeFalsy()
    })

    it('has no clear button when it is enabled but the input is an empty string', () => {
      const wrapper = getShallowWrapper({
        clearButtonEnabled: true,
        modelValue: ''
      })

      expect(wrapper.find(selectors.clearInputButton).exists()).toBeFalsy()
    })

    it('has no clear button when it is enabled but the input is null', () => {
      const wrapper = getShallowWrapper({
        clearButtonEnabled: true,
        modelValue: null
      })

      expect(wrapper.find(selectors.clearInputButton).exists()).toBeFalsy()
    })

    it('has a clear button when it is enabled and the input contains content', () => {
      const wrapper = getShallowWrapper({
        clearButtonEnabled: true,
        modelValue: 'non-empty-value'
      })

      const btn = wrapper.find(selectors.clearInputButton)
      expect(btn.exists()).toBeTruthy()
    })

    it('clears the content on click', async () => {
      const wrapper = getMountedWrapper({
        props: {
          ...defaultProps,
          clearButtonEnabled: true,
          modelValue: 'non-empty-value'
        },
        attachTo: document.body
      })

      const btn = wrapper.find(selectors.clearInputButton)
      const input = wrapper.find(selectors.inputField)

      await btn.trigger('click')

      // value as data is supposed to be `null`
      expect(wrapper.emitted('update:modelValue')[0][0]).toEqual(null)
      // value in DOM would be the empty string if two way binding was used
      // by just passing in the value it should remain unchanged
      expect((input.element as HTMLInputElement).value).toEqual('non-empty-value')
      expect(document.activeElement.id).toBe(input.element.id)
    })
  })
})
