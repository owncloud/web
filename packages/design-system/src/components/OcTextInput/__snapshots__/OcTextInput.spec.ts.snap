// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OcTextInput > password input field > password policy > displays error state if password does not match criteria 1`] = `
"<div class=""><label class="oc-label" for="oc-textinput-24"></label>
  <div class="oc-position-relative">
    <!--v-if-->
    <div class="oc-text-input-password-wrapper"><input id="oc-textinput-24" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="password"> <button type="button" aria-label="Show password" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-input-show-password-toggle oc-px-s oc-background-default">
        <!--v-if-->
        <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
      </button> <button type="button" aria-label="Copy password" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-input-copy-password-button oc-px-s oc-background-default">
        <!--v-if-->
        <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
      </button>
      <!--v-if-->
    </div>
    <portal to="app.design-system.password-policy">
      <div class="oc-text-small"><span>Please enter a password that meets the following criteria:</span>
        <div class="oc-flex oc-text-input-password-policy-rule-wrapper">
          <div class="oc-flex oc-flex-middle oc-text-input-password-policy-rule"><span class="oc-icon oc-icon-s oc-icon-danger oc-mr-xs"><!----></span> <span class="oc-text-input-danger">8+ letters</span>
            <!--v-if-->
          </div>
        </div>
      </div>
    </portal>
    <!--v-if-->
  </div>
  <!--v-if-->
  <portal-target name="app.design-system.password-policy"></portal-target>
</div>"
`;

exports[`OcTextInput > password input field > password policy > displays success state if password matches criteria 1`] = `
"<div class=""><label class="oc-label" for="oc-textinput-25"></label>
  <div class="oc-position-relative">
    <!--v-if-->
    <div class="oc-text-input-password-wrapper"><input id="oc-textinput-25" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="password"> <button type="button" aria-label="Show password" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-input-show-password-toggle oc-px-s oc-background-default">
        <!--v-if-->
        <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
      </button> <button type="button" aria-label="Copy password" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-input-copy-password-button oc-px-s oc-background-default">
        <!--v-if-->
        <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
      </button>
      <!--v-if-->
    </div>
    <portal to="app.design-system.password-policy">
      <div class="oc-text-small"><span>Please enter a password that meets the following criteria:</span>
        <div class="oc-flex oc-text-input-password-policy-rule-wrapper">
          <div class="oc-flex oc-flex-middle oc-text-input-password-policy-rule"><span class="oc-icon oc-icon-s oc-icon-success oc-mr-xs"><!----></span> <span class="oc-text-input-success">8+ letters</span>
            <!--v-if-->
          </div>
        </div>
      </div>
    </portal>
    <!--v-if-->
  </div>
  <!--v-if-->
  <portal-target name="app.design-system.password-policy"></portal-target>
</div>"
`;
